blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCacheCloneReadSaveRequest.java

```diff
@@ -0,0 +1,247 @@
+            return fs.invoke(new GitSCMFileSystem.FSFunction<byte[]>() {
+                @Override
+                public byte[] invoke(Repository repository) throws IOException, InterruptedException {
+                    Git activeRepo = getActiveRepository(repository);
+                    Repository repo = activeRepo.getRepository();
+                    File repoDir = repo.getDirectory().getParentFile();
+                    try {
+                        File f = new File(repoDir, filePath);
+                        if (f.canRead()) {
+                            return IOUtils.toByteArray(new FileInputStream(f));
+                        }
+                        return null;
+                    } finally {
+                        FileUtils.deleteDirectory(repoDir);
+                    }
+                }
+            });
+        } catch (InterruptedException ex) {
+            throw new ServiceException.UnexpectedErrorException("Unable to read " + filePath, ex);
+        }
+    }
+
+    @Override
+    void save() throws IOException {
+        try {
+            GitSCMFileSystem fs = getFilesystem();
+            fs.invoke(new GitSCMFileSystem.FSFunction<Void>() {
+                @Override
+                public Void invoke(Repository repository) throws IOException, InterruptedException {
+                    Git activeRepo = getActiveRepository(repository);
+                    Repository repo = activeRepo.getRepository();
+                    File repoDir = repo.getDirectory().getParentFile();
+                    log.fine("Repo cloned to: " + repoDir.getCanonicalPath());
+                    try {
+                        File f = new File(repoDir, filePath);
+                        if (!f.exists() || f.canWrite()) {
+                            try (Writer w = new OutputStreamWriter(new FileOutputStream(f), "utf-8")) {
+                                w.write(new String(contents, "utf-8"));
+                            }
+
+                            try {
+                                AddCommand add = activeRepo.add();
+                                add.addFilepattern(filePath);
+                                add.call();
+
+                                CommitCommand commit = activeRepo.commit();
+                                commit.setMessage(commitMessage);
+                                commit.call();
+
+                                StandardCredentials credential = null;
+                                if (GitUtils.isSshUrl(gitSource.getRemote())) {
+                                    // Get committer info and credentials
+                                    User user = User.current();
+                                    if (user == null) {
+                                        throw new ServiceException.UnauthorizedException("Not authenticated");
+                                    }
+                                    credential = UserSSHKeyManager.getOrCreate(user);
```

Comment:
@kzantow `UserSSHKeyManager.getOrCreate(user)` doesn't seem right here in case where credential id has been deleted or has not been setup then whats the point of creating a new ssh key pair/credential when the repo might not be setup to allow this new public key?

Reply:
Maybe doing `UserSSHKeyManager.get(user)` and then throw 428 if credential is null?

Reply:
Well, presumably the user has viewed the UI, in which case the key would have been created and displayed, I don't see much reason to modify the behavior here.

Reply:
Umm, this is API and can be curled. API should return consistent error code, in this case that's expected behavior vs failing with some 500 error code.

Reply:
This will attempt to connect with an SSH key if using an SSH URL. Whether the user has done something with it or not is irrelevant. If connecting fails, it'll return 428, not some 500 error.

Reply:
Great, but where does it throwing 428. I checked GitClient.clone() or its other method, they seem to throw RuntimeException. Other places in the code I see these are caught and 500 is returned.